### ISSUE TITLE
Relationship between ARRIVE,SERVE,DONE,WAIT,LEAVE

### ISSUE THREAD

**Issue by johnnythesnake12** [nus-cs2030/2122-s2#390]
Hi , 
Could anyone tell me how i should design the relationship between ARRIVE, SERVE, DONE, WAIT, LEAVE.

**Comment by ngnevan1** [comment]
You could start by using the flow from the task description
ARRIVE → SERVE → DONE
ARRIVE → WAIT → SERVE → DONE
ARRIVE → LEAVE

Every event starts from Arrive. Each Arrive event would return then check the state of the shop and return the next appropriate event and an updated shop. For the Done and Leave events, they would return Optional.empty() and the updated shop

You might have to add intermediate events for the later levels

**Comment by bryankwe** [comment]
This was what my execute() produced for each type of event: (Try is for the upper levels)
Arrive --> Wait / Serve / Leave
Wait --> Try
Try --> Try / Serve
Serve --> Done

Off course, there are other implementations which I have seen others use. It is really up to you and how you decide to tackle the project

**Comment by Guanjung27** [comment]
make use of the logic as described by the question 
what i used was this:
ARRIVE → SERVE → DONE
ARRIVE → WAIT → SERVE → DONE
ARRIVE → LEAVE
there is probably another class later on , but this is the general idea

**Comment by yongchein-o** [comment]
Yeah and done should return Optional.empty()
I initailly returned serve from done however, that led to cyclic dependency

**Comment by weihan10** [comment]
Yup, you can just follow from the task description. There is absolutely no need to create additional try classes in my opinion, and also your code will look cleaner without such additons too.

**Comment by johnnythesnake12** [comment]
Thank you all for your suggestions :D
